content/blog/2019-11-13-loop-perforation/index.md

```diff
@@ -0,0 +1,420 @@
+Rather than using the absolute scale of the correct answer as our measuring stick, we effectively provide the variance as an input.
+Assuming that the final distance measure is normally distributed (which is more and more appropriate the more sums and multiplications of random variables are involved) with variance $\sigma^2/2$, then the probability that the observed variable lies within a distance $d$ of the correct one is exactly the [_error function_](https://en.wikipedia.org/wiki/Error_function):
+
+\[ \text{erf}(x) := \frac{1}{\sqrt{\pi}} \int_{-x}^x e^{-t^2} \mathrm{d}t \]
+
+
+This can be scaled to a specific variance $\sigma^2/2$ by dividing $t$ by $\sigma^2$.
+Effectively, this gives us a way of turning scalar distances into variance-parameterized error metrics.
+For these reasons, we enter a total accuracy score for each error metric that we collect, as well as for each of many variances.
+We then use all of these metricscass to compute the Pareto frontier.
+
+These error metrics are more finicky than they might appear for real programs.
+Consider a Sobel image filter for edge detection.
+We use an implementation of this filter from the [ACCEPT benchmarks][] for approximate computing.
+
+Running the standard implementation on this image of cute buffalo:
+
+<img src="sobel-input.png" width=500/>
+
+Gives the following correct output:
+
+<img src="sobel-output.png" width=500/>
+
+Running our driver with a default configuration perforates the loops in this application at the following rates:
+
+```
+{
+    \"load_image_data\": {\"%81,%84,%85,%97,%98,%88,%95\": 5, \"%85,%88,%95\": 5},
+    \"save_image_data\": {\"%11,%14,%15,%28,%29,%18,%26\": 1, \"%15,%18,%26\": 1},
+    \"sobel_filtering\": {\"%74,%77,%78,%88,%89,%81,%86\": 5, \"%78,%81,%86\": 5}
+}
+```
+
+The driver chooses this configuration because one of the error metrics, `"l2_100000"`, only has an error of 0.185.
+However the resulting image looks...very bad.
+
+
+<img src="sobel-perforated-default.png" width=500/>
+
+With a little design space exploration, we can try and find other metrics that produce less psychedelic results.
+The following result only considers the `l1_10000` metric and allows an error of 0.6.
+
+<img src="sobel-perforated-l1-10000-0.6.png" width=500/>
+
+This image is _kind of_ on the right track, in that we can clearly see some of the edges in the original image (just shifted and stretched to be almost useless).
+Remarkably, the above image was the closest to correct we could find with any loop perforation at all via a casual design space exploration.
+However, this is clearly **not** an acceptable result for edge detection!
+
+Ultimately, this example illustrates that finding the right error metric might be both difficult and counterintuitive—to get a better image that gained any speedup, we had to select a higher error rate, just for a different metric.
```

Comment:
c.f. "No more sadness about the imperfection of quality metrics."
https://www.cs.cornell.edu/~asampson/blog/closedproblems.html#closed-domains-for-approximation